src/lib/mf.rs

```diff
@@ -509,16 +476,398 @@ fn rank_cnds<TokId: Clone + Copy + Debug + TryFrom<usize> + TryInto<usize> + Par
         .collect::<Vec<Vec<ParseRepair>>>()
 }
 
+pub(crate) struct Dist {
+    terms_len: usize,
+    table: Vec<u64>
+}
+
+impl Dist {
+    pub(crate) fn new<F>(grm: &YaccGrammar, sgraph: &StateGraph, stable: &StateTable, term_cost: F) -> Dist
+              where F: Fn(TIdx) -> u64
+    {
+        // This is an extension of dist from the KimYi paper: it also takes into account reductions
+        // and gotos in the distances it reports back. Note that it is conservative, sometimes
+        // undercounting the distance. Consider the KimYi paper, Figure 10: from state 0 it is
+        // clearly the case that the distance to ')' is 2 since we would need to encounter the
+        // tokens "(", {"a", "b"} before encountering a ')'. However, this algorithm gives the
+        // distance as 1. This is because of state 0 is a reduction target of state 3, but state 3
+        // also has a reduction target of state 5: thus state state 0 ends up with the minimum
+        // distances of itself and state 5. Put another way, state 0 and state 5 end up with the
+        // same set of distances, even though a more clever algorithm could work out that this
+        // needn't be so. That's for another day...
+
+        let terms_len = grm.terms_len();
+        let states_len = sgraph.all_states_len();
+        let sengen = grm.sentence_generator(&term_cost);
+        let mut table = Vec::new();
+        table.resize(states_len * terms_len, u64::max_value());
+        loop {
+            let mut chgd = false; // Has anything changed?
```

Comment:
This is a nitpick, but does it really hurt so much to write those 3 addtional characters of `changed`? Did you know you could save even more keypresses by writing `tbl` and `egds` instead of `table` and `edges`? ;-)

Reply:
I used to love long variable names, but I think I was wrong: the best code I've read has short variable names like the above. Yes, I still find it a bit weird sometimes, and I still lapse into longer names... but I'm trying :)